### ISSUE TITLE
MMP loss metrics are systematically incorrect (off-by-one, bad guard, ETX inconsistency)

### ISSUE THREAD

**Issue by v0l** [jmcorgan/fips#14]
## Summary

Three bugs in the MMP (Mesh Management Protocol) metric calculation code cause packet loss to be reported incorrectly. These were identified by cross-referencing both sides of the connection in issue #13 — kieran-x reported 40–50% packet loss while the remote end (`fips-v0l-io`) saw 0% loss on the same link throughout.

The root cause is not network packet loss but measurement artifacts in `src/mmp/metrics.rs` and the reverse-ratio calculation in the node handlers.

---

## Bug C (Primary) — Off-by-one in `counter_span` inflates loss at low frame rates

**File:** `src/mmp/metrics.rs`

Loss is computed as:
```rust
let counter_span = rr.highest_counter.saturating_sub(self.prev_rr_highest_counter);
let delivery_ratio = packets_delta as f64 / counter_span as f64;
let loss = 1.0 - delivery_ratio;
```

`highest_counter` tracks the highest AEAD frame counter seen. If the previous report had `highest_counter = 5` and the current report has `highest_counter = 7`, then `counter_span = 2`. But the frames potentially sent in that window are counters 5, 6, and 7 — **3 frames**, not 2. The frame at `prev_rr_highest_counter` is included in the previous interval's `packets_delta` but excluded from the current interval's `counter_span`, making the denominator consistently 1 too small.

At high frame rates this is negligible (1/1000 = 0.1% error). At low frame rates it is severe: a session with 2 frames/interval has `counter_span = 1`, so even if both frames are received `packets_delta/counter_span = 2/1` which clamps to 1.0 — but if one frame is a control message not counted in `packets_delta`, the result is `1/1 = 1.0` (fine). The real damage is when the session is lightly loaded with ~3 frames/interval and actual delivery is perfect: `counter_span` reports 2, `packets_delta` is 2, loss = 0% — but add one missed count and it's instantly 50%.

**Fix:** Change `counter_span` to `rr.highest_counter.saturating_sub(self.prev_rr_highest_counter) + 1`, or redesign `highest_counter` to track the next expected counter (i.e. increment it to `counter + 1` on receive) so the subtraction naturally gives the correct span.

---

## Bug B — `prev_rr_highest_counter > 0` guard skips loss for the entire first measurement interval

**File:** `src/mmp/metrics.rs`

```rust
if self.prev_rr_highest_counter > 0 {
    // loss computation
}
```

FSP and FMP counters start at 0. The first frame received has counter = 0, so `highest_counter` stays 0 after the first ReceiverReport. The guard `> 0` then fails, and the entire first measurement interval is silently skipped. `prev_rr_highest_counter` is then saved as 0 again. Only when a subsequent report has `highest_counter > 0` does loss computation finally run — but even then, the span covers from 0 to the current value, meaning the very first report's data is double-counted in the denominator.

**Fix:** Replace the `> 0` guard with a dedicated `has_prev_rr: bool` field, set to `true` after the first ReceiverReport is processed. This correctly handles the counter-0 case.

---

## Bug A — `delivery_ratio_reverse` uses a lifetime cumulative ratio instead of an interval delta

**Files:** `src/node/handlers/mmp.rs`, `src/node/handlers/session.rs`

```rust
let our_recv_packets = mmp.receiver.cumulative_packets_recv();
let peer_highest = mmp.receiver.highest_counter();
if peer_highest > 0 {
    let reverse_ratio = (our_recv_packets as f64) / (peer_highest as f64);
    mmp.metrics.set_delivery_ratio_reverse(reverse_ratio);
}
```

The forward delivery ratio (`delivery_ratio_forward`) is computed as a **per-interval delta**: `packets_recv_this_interval / counters_advanced_this_interval`. The reverse ratio is computed as a **lifetime cumulative**: `total_packets_ever_received / highest_counter_ever_seen`. These measure fundamentally different things and are inconsistent with each other, making ETX unreliable — particularly on young sessions where the ratio is volatile.

This is why teardown logs show `etx=1.28` on a connection that reported `loss=0.0%` throughout its lifetime.

**Fix:** Compute `delivery_ratio_reverse` using per-interval deltas, the same way `delivery_ratio_forward` is computed. Store `prev_rr_cumulative_recv` and `prev_rr_highest_counter` together and compute both ratios from the same interval window.

---

## Observed Symptoms (from issue #13)

- `kieran-x` reported MMP session loss of 7.9% → 14.1% → 24.5% → 40–50% on a session to `fips-v0l-io`
- `fips-v0l-io` reported **0.0% loss** at every measurement point on the same link
- RTT was stable at ~30ms throughout with no actual network packet loss (confirmed by independent packet capture)
- Session teardown on VM side showed `etx=1.00` with `loss=0.0%`; client side teardown showed elevated ETX despite same 0% loss reading

All three symptoms are consistent with the bugs above.

**Comment by jmcorgan** [comment]
**Bug A (reverse ratio lifetime cumulative) — fixed in 1020f90**

`update_reverse_delivery()` now tracks previous snapshot state (`prev_reverse_packets`, `prev_reverse_highest`) and computes per-interval deltas, matching the forward ratio's approach. The `has_prev_reverse` flag handles baseline establishment cleanly (no `> 0` guard needed).

**Comment by jmcorgan** [comment]
As for Bug C, I don't think this is actually a bug. The `+1` reasoning double-counts the boundary frame. Both `highest_counter` and `cumulative_packets_recv` are snapshotted from the same ReceiverReport, so when `prev_rr_highest_counter = 5`, frame 5 is already included in that report's `cumulative_packets_recv`. When the next report arrives with `highest_counter = 7`, `packets_delta` counts only frames received after the snapshot (frames 6 and 7), and `counter_span = 7 - 5 = 2` correctly matches: two frames could have been sent in that interval.

**Comment by jmcorgan** [comment]
As for Bug B, the `> 0` guard is technically wrong since counters start at 0, but the only interval that gets skipped is the first one, which needs to be skipped anyway to establish a baseline for delta computation. Fixed in 6e03add with a proper `has_prev_rr` flag, but it's not causing the symptoms described in issue #13.

**Comment by jmcorgan** [comment]
Merged to master.
